File: ReactCommon/react/renderer/mapbuffer/MapBuffer.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "MapBuffer.h"

using namespace facebook::react;

namespace facebook {
namespace react {

// TODO T83483191: Extend MapBuffer C++ implementation to support basic random
// access
MapBuffer::MapBuffer(std::vector<uint8_t> data) : bytes_(std::move(data)) {
  count_ = 0;
  memcpy(
      reinterpret_cast<uint8_t *>(&count_),
      bytes_.data() + HEADER_COUNT_OFFSET,
      UINT16_SIZE);

  // TODO T83483191: extract memcpy calls into an inline function to simplify
  // the code
  int32_t dataSize;
  memcpy(
      reinterpret_cast<uint8_t *>(&dataSize),
      bytes_.data() + HEADER_BUFFER_SIZE_OFFSET,
      INT_SIZE);

  if (dataSize != bytes_.size()) {
    LOG(ERROR) << "Error: Data size does not match, expected " << dataSize
               << " found: " << bytes_.size();
    abort();
  }
}

int32_t MapBuffer::getInt(Key key) const {
  int32_t value = 0;
  memcpy(
      reinterpret_cast<uint8_t *>(&value),
      bytes_.data() + getValueOffset(key),
      INT_SIZE);
  return value;
}

bool MapBuffer::getBool(Key key) const {
  return getInt(key) != 0;
}

double MapBuffer::getDouble(Key key) const {
  // TODO T83483191: extract this code into a "template method" and reuse it for
  // other types
  double value = 0;
  memcpy(
      reinterpret_cast<uint8_t *>(&value),
      bytes_.data() + getValueOffset(key),
      DOUBLE_SIZE);
  return value;
}

int32_t MapBuffer::getDynamicDataOffset() const {
  // The begininig of dynamic data can be calculated as the offset of the next
  // key in the map
  return getKeyOffset(count_);
}

std::string MapBuffer::getString(Key key) const {
  // TODO T83483191:Add checks to verify that offsets are under the boundaries
  // of the map buffer
  int32_t dynamicDataOffset = getDynamicDataOffset();
  int32_t stringLength = 0;
  int32_t offset = getInt(key);
  memcpy(
      reinterpret_cast<uint8_t *>(&stringLength),
      bytes_.data() + dynamicDataOffset + offset,
      INT_SIZE);

  char *value = new char[stringLength];

  memcpy(
      reinterpret_cast<char *>(value),
      bytes_.data() + dynamicDataOffset + offset + INT_SIZE,
      stringLength);

  return std::string(value, 0, stringLength);
}

MapBuffer MapBuffer::getMapBuffer(Key key) const {
  // TODO T83483191: Add checks to verify that offsets are under the boundaries
  // of the map buffer
  int32_t dynamicDataOffset = getDynamicDataOffset();

  int32_t mapBufferLength = 0;
  int32_t offset = getInt(key);
  memcpy(
      reinterpret_cast<uint8_t *>(&mapBufferLength),
      bytes_.data() + dynamicDataOffset + offset,
      INT_SIZE);

  std::vector<uint8_t> value(mapBufferLength);

  memcpy(
      value.data(),
      bytes_.data() + dynamicDataOffset + offset + INT_SIZE,
      mapBufferLength);

  return MapBuffer(std::move(value));
}

bool MapBuffer::isNull(Key key) const {
  return getInt(key) == NULL_VALUE;
}

int32_t MapBuffer::getBufferSize() const {
  return bytes_.size();
}

void MapBuffer::copy(uint8_t *output) const {
  memcpy(output, bytes_.data(), bytes_.size());
}

uint16_t MapBuffer::getCount() const {
  uint16_t size = 0;

  memcpy(
      reinterpret_cast<uint8_t *>(&size),
      bytes_.data() + HEADER_COUNT_OFFSET,
      UINT16_SIZE);

  return size;
}

} // namespace react
} // namespace facebook

File: ReactAndroid/src/main/java/com/facebook/react/common/mapbuffer/MapBuffer.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.common.mapbuffer

/**
 * MapBuffer is an optimized sparse array format for transferring props-like data between C++ and
 * JNI. It is designed to:
 * - be compact to optimize space when sparse (sparse is the common case).
 * - be accessible through JNI with zero/minimal copying.
 * - work recursively for nested maps/arrays.
 * - support dynamic types that map to JSON.
 * - have minimal APK size and build time impact.
 *
 * See <react/renderer/mapbuffer/MapBuffer.h> for more information and native implementation.
 *
 * Limitations:
 * - Keys are usually sized as 2 bytes, with each buffer supporting up to 65536 entries as a result.
 * - O(log(N)) random key access for native buffers due to selected structure. Faster access can be
 * achieved by retrieving [MapBuffer.Entry] with [entryAt] on known offsets.
 */
interface MapBuffer : Iterable<MapBuffer.Entry> {
  companion object {
    /**
     * Key are represented as 2 byte values, and typed as Int for ease of access. The serialization
     * format only allows to store [UShort] values.
     */
    internal val KEY_RANGE = IntRange(UShort.MIN_VALUE.toInt(), UShort.MAX_VALUE.toInt())
  }

  /**
   * Data types supported by [MapBuffer]. Keep in sync with definition in
   * `<react/renderer/mapbuffer/MapBuffer.h>`, as enum serialization relies on correct order.
   */
  enum class DataType {
    BOOL,
    INT,
    DOUBLE,
    STRING,
    MAP
  }

  /**
   * Number of elements inserted into current MapBuffer.
   * @return number of elements in the [MapBuffer]
   */
  val count: Int

  /**
   * Checks whether entry for given key exists in MapBuffer.
   * @param key key to lookup the entry
   * @return whether entry for the given key exists in the MapBuffer.
   */
  fun contains(key: Int): Boolean

  /**
   * Provides offset of the key to use for [entryAt], for cases when offset is not statically known
   * but can be cached.
   * @param key key to lookup offset for
   * @return offset for the given key to be used for entry access, -1 if key wasn't found.
   */
  fun getKeyOffset(key: Int): Int

  /**
   * Provides parsed access to a MapBuffer without additional lookups for provided offset.
   * @param offset offset of entry in the MapBuffer structure. Can be looked up for known keys with
   * [getKeyOffset].
   * @return parsed entry for structured access for given offset
   */
  fun entryAt(offset: Int): MapBuffer.Entry

  /**
   * Provides parsed [DataType] annotation associated with the given key.
   * @param key key to lookup type for
   * @return data type of the given key.
   * @throws IllegalArgumentException if the key doesn't exist
   */
  fun getType(key: Int): DataType

  /**
   * Provides parsed [Boolean] value if the entry for given key exists with [DataType.BOOL] type
   * @param key key to lookup [Boolean] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getBoolean(key: Int): Boolean

  /**
   * Provides parsed [Int] value if the entry for given key exists with [DataType.INT] type
   * @param key key to lookup [Int] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getInt(key: Int): Int

  /**
   * Provides parsed [Double] value if the entry for given key exists with [DataType.DOUBLE] type
   * @param key key to lookup [Double] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getDouble(key: Int): Double

  /**
   * Provides parsed [String] value if the entry for given key exists with [DataType.STRING] type
   * @param key key to lookup [String] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getString(key: Int): String

  /**
   * Provides parsed [MapBuffer] value if the entry for given key exists with [DataType.MAP] type
   * @param key key to lookup [MapBuffer] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getMapBuffer(key: Int): MapBuffer

  /**
   * Provides parsed [List<MapBuffer>] value if the entry for given key exists with [DataType.MAP]
   * type
   * @param key key to lookup [List<MapBuffer>] value for
   * @return value associated with the requested key
   * @throws IllegalArgumentException if the key doesn't exist
   * @throws IllegalStateException if the data type doesn't match
   */
  fun getMapBufferList(key: Int): List<MapBuffer>

  /** Iterable entry representing parsed MapBuffer values */
  interface Entry {
    /**
     * Key of the given entry. Usually represented as 2 byte unsigned integer with the value range
     * of [0,65536)
     */
    val key: Int

    /** Parsed [DataType] of the entry */
    val type: DataType

    /**
     * Entry value represented as [Boolean]
     * @throws IllegalStateException if the data type doesn't match [DataType.BOOL]
     */
    val booleanValue: Boolean

    /**
     * Entry value represented as [Int]
     * @throws IllegalStateException if the data type doesn't match [DataType.INT]
     */
    val intValue: Int

    /**
     * Entry value represented as [Double]
     * @throws IllegalStateException if the data type doesn't match [DataType.DOUBLE]
     */
    val doubleValue: Double

    /**
     * Entry value represented as [String]
     * @throws IllegalStateException if the data type doesn't match [DataType.STRING]
     */
    val stringValue: String

    /**
     * Entry value represented as [MapBuffer]
     * @throws IllegalStateException if the data type doesn't match [DataType.MAP]
     */
    val mapBufferValue: MapBuffer
  }
}
